File: ruby/lib/remitmd/x402_paywall.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"
require "base64"

module Remitmd
  # x402 paywall for service providers - gate HTTP endpoints behind payments.
  #
  # Providers use this class to:
  # - Return HTTP 402 responses with properly formatted PAYMENT-REQUIRED headers
  # - Verify incoming PAYMENT-SIGNATURE headers against the remit.md facilitator
  #
  # @example Rack middleware
  #   paywall = Remitmd::X402Paywall.new(
  #     wallet_address: "0xYourProviderWallet",
  #     amount_usdc: 0.001,
  #     network: "eip155:84532",
  #     asset: "0x2d846325766921935f37d5b4478196d3ef93707c"
  #   )
  #   use paywall.rack_middleware
  #
  class X402Paywall
    # @param wallet_address [String] provider's checksummed Ethereum address (the payTo field)
    # @param amount_usdc [Float] price per request in USDC (e.g. 0.001)
    # @param network [String] CAIP-2 network string (e.g. "eip155:84532")
    # @param asset [String] USDC contract address on the target network
    # @param facilitator_url [String] base URL of the remit.md facilitator
    # @param facilitator_token [String] bearer JWT for authenticating calls to /api/v1/x402/verify
    # @param max_timeout_seconds [Integer] how long the payment authorization remains valid
    # @param resource [String, nil] V2 - URL or path of the resource being protected
    # @param description [String, nil] V2 - human-readable description
    # @param mime_type [String, nil] V2 - MIME type of the resource
    def initialize( # rubocop:disable Metrics/ParameterLists
      wallet_address:,
      amount_usdc:,
      network:,
      asset:,
      facilitator_url: "https://remit.md",
      facilitator_token: "",
      max_timeout_seconds: 60,
      resource: nil,
      description: nil,
      mime_type: nil
    )
      @wallet_address      = wallet_address
      @amount_base_units   = (amount_usdc * 1_000_000).round.to_s
      @network             = network
      @asset               = asset
      @facilitator_url     = facilitator_url.chomp("/")
      @facilitator_token   = facilitator_token
      @max_timeout_seconds = max_timeout_seconds
      @resource            = resource
      @description         = description
      @mime_type           = mime_type
    end

    # Return the base64-encoded JSON PAYMENT-REQUIRED header value.
    # @return [String]
    def payment_required_header
      payload = {
        scheme:            "exact",
        network:           @network,
        amount:            @amount_base_units,
        asset:             @asset,
        payTo:             @wallet_address,
        maxTimeoutSeconds: @max_timeout_seconds,
      }
      payload[:resource]    = @resource    if @resource
      payload[:description] = @description if @description
      payload[:mimeType]    = @mime_type   if @mime_type
      Base64.strict_encode64(JSON.generate(payload))
    end

    # Check whether a PAYMENT-SIGNATURE header represents a valid payment.
    # Calls the remit.md facilitator's /api/v1/x402/verify endpoint.
    #
    # @param payment_sig [String, nil] the raw header value (base64 JSON), or nil if absent
    # @return [Hash] { is_valid: true/false, invalid_reason: String or nil }
    def check(payment_sig)
      return { is_valid: false } unless payment_sig

      payment_payload = begin
        JSON.parse(Base64.decode64(payment_sig))
      rescue JSON::ParserError
        return { is_valid: false, invalid_reason: "INVALID_PAYLOAD" }
      end

      body = {
        paymentPayload:  payment_payload,
        paymentRequired: payment_required_object,
      }

      uri = URI("#{@facilitator_url}/api/v1/x402/verify")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = uri.scheme == "https"
      http.read_timeout = 10

      req = Net::HTTP::Post.new(uri.path)
      req["Content-Type"] = "application/json"
      req["Authorization"] = "Bearer #{@facilitator_token}" unless @facilitator_token.empty?
      req.body = JSON.generate(body)

      begin
        resp = http.request(req)
        unless resp.is_a?(Net::HTTPSuccess)
          return { is_valid: false, invalid_reason: "FACILITATOR_ERROR" }
        end

        data = JSON.parse(resp.body)
      rescue StandardError
        return { is_valid: false, invalid_reason: "FACILITATOR_ERROR" }
      end

      {
        is_valid:       data["isValid"] == true,
        invalid_reason: data["invalidReason"],
      }
    end

    # Rack middleware adapter.
    #
    # @example
    #   use paywall.rack_middleware
    #
    # @return [Class] a Rack middleware class
    def rack_middleware
      paywall = self
      Class.new do
        define_method(:initialize) do |app|
          @app     = app
          @paywall = paywall
        end

        define_method(:call) do |env|
          payment_sig = env["HTTP_PAYMENT_SIGNATURE"]
          result = @paywall.check(payment_sig)

          unless result[:is_valid]
            headers = {
              "Content-Type"     => "application/json",
              "PAYMENT-REQUIRED" => @paywall.payment_required_header,
            }
            body = JSON.generate({
              error:         "Payment required",
              invalidReason: result[:invalid_reason],
            })
            return [402, headers, [body]]
          end

          @app.call(env)
        end
      end
    end

    private

    def payment_required_object
      {
        scheme:            "exact",
        network:           @network,
        amount:            @amount_base_units,
        asset:             @asset,
        payTo:             @wallet_address,
        maxTimeoutSeconds: @max_timeout_seconds,
      }
    end
  end
end

File: dotnet/RemitMd/X402Paywall.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace RemitMd;

/// <summary>Result of <see cref="X402Paywall.CheckAsync"/>.</summary>
public sealed class CheckResult
{
    /// <summary>Whether the payment signature is valid.</summary>
    [JsonPropertyName("isValid")]
    public bool IsValid { get; set; }

    /// <summary>Reason the payment was rejected (when <see cref="IsValid"/> is false).</summary>
    [JsonPropertyName("invalidReason")]
    public string? InvalidReason { get; set; }
}

/// <summary>
/// x402 paywall for service providers.
///
/// Providers use this class to:
/// <list type="bullet">
///   <item>Return HTTP 402 responses with properly formatted <c>PAYMENT-REQUIRED</c> headers</item>
///   <item>Verify incoming <c>PAYMENT-SIGNATURE</c> headers against the remit.md facilitator</item>
/// </list>
/// </summary>
public sealed class X402Paywall : IDisposable
{
    private readonly string _walletAddress;
    private readonly string _amountBaseUnits;
    private readonly string _network;
    private readonly string _asset;
    private readonly string _facilitatorUrl;
    private readonly string _facilitatorToken;
    private readonly int _maxTimeoutSeconds;
    private readonly string? _resource;
    private readonly string? _description;
    private readonly string? _mimeType;
    private readonly HttpClient _http;

    /// <summary>
    /// Creates an X402 paywall for gating endpoints behind USDC payments.
    /// </summary>
    /// <param name="walletAddress">Provider's checksummed Ethereum address (the payTo field).</param>
    /// <param name="amountUsdc">Price per request in USDC (e.g. 0.001).</param>
    /// <param name="network">CAIP-2 network string (e.g. "eip155:84532" for Base Sepolia).</param>
    /// <param name="asset">USDC contract address on the target network.</param>
    /// <param name="facilitatorUrl">Base URL of the remit.md facilitator (default: "https://remit.md").</param>
    /// <param name="facilitatorToken">Bearer JWT for authenticating calls to /api/v1/x402/verify.</param>
    /// <param name="maxTimeoutSeconds">How long the payment authorization remains valid (default: 60).</param>
    /// <param name="resource">V2 - URL or path of the resource being protected.</param>
    /// <param name="description">V2 - Human-readable description of what the payment is for.</param>
    /// <param name="mimeType">V2 - MIME type of the resource (e.g. "application/json").</param>
    public X402Paywall(
        string walletAddress,
        decimal amountUsdc,
        string network,
        string asset,
        string facilitatorUrl = "https://remit.md",
        string facilitatorToken = "",
        int maxTimeoutSeconds = 60,
        string? resource = null,
        string? description = null,
        string? mimeType = null)
    {
        _walletAddress = walletAddress;
        _amountBaseUnits = ((long)Math.Round(amountUsdc * 1_000_000m)).ToString();
        _network = network;
        _asset = asset;
        _facilitatorUrl = facilitatorUrl.TrimEnd('/');
        _facilitatorToken = facilitatorToken;
        _maxTimeoutSeconds = maxTimeoutSeconds;
        _resource = resource;
        _description = description;
        _mimeType = mimeType;
        _http = new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
    }

    /// <summary>Return the base64-encoded JSON <c>PAYMENT-REQUIRED</c> header value.</summary>
    public string PaymentRequiredHeader()
    {
        var payload = new Dictionary<string, object>
        {
            ["scheme"] = "exact",
            ["network"] = _network,
            ["amount"] = _amountBaseUnits,
            ["asset"] = _asset,
            ["payTo"] = _walletAddress,
            ["maxTimeoutSeconds"] = _maxTimeoutSeconds,
        };
        if (_resource is not null) payload["resource"] = _resource;
        if (_description is not null) payload["description"] = _description;
        if (_mimeType is not null) payload["mimeType"] = _mimeType;

        var json = JsonSerializer.Serialize(payload);
        return Convert.ToBase64String(Encoding.UTF8.GetBytes(json));
    }

    /// <summary>
    /// Check whether a <c>PAYMENT-SIGNATURE</c> header represents a valid payment.
    /// Calls the remit.md facilitator's <c>/api/v1/x402/verify</c> endpoint.
    /// </summary>
    /// <param name="paymentSig">The raw header value (base64 JSON), or null if absent.</param>
    /// <returns><see cref="CheckResult"/> indicating validity.</returns>
    public async Task<CheckResult> CheckAsync(string? paymentSig, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(paymentSig))
            return new CheckResult { IsValid = false };

        object paymentPayload;
        try
        {
            var decoded = Encoding.UTF8.GetString(Convert.FromBase64String(paymentSig));
            paymentPayload = JsonSerializer.Deserialize<object>(decoded)
                ?? throw new InvalidOperationException();
        }
        catch
        {
            return new CheckResult { IsValid = false, InvalidReason = "INVALID_PAYLOAD" };
        }

        var body = new
        {
            paymentPayload,
            paymentRequired = new
            {
                scheme = "exact",
                network = _network,
                amount = _amountBaseUnits,
                asset = _asset,
                payTo = _walletAddress,
                maxTimeoutSeconds = _maxTimeoutSeconds,
            },
        };

        try
        {
            var request = new HttpRequestMessage(HttpMethod.Post,
                $"{_facilitatorUrl}/api/v1/x402/verify");
            request.Content = new StringContent(
                JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");
            if (!string.IsNullOrEmpty(_facilitatorToken))
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _facilitatorToken);

            var resp = await _http.SendAsync(request, ct);
            if (!resp.IsSuccessStatusCode)
                return new CheckResult { IsValid = false, InvalidReason = "FACILITATOR_ERROR" };

            var resultJson = await resp.Content.ReadAsStringAsync(ct);
            var result = JsonSerializer.Deserialize<CheckResult>(resultJson);
            return result ?? new CheckResult { IsValid = false, InvalidReason = "FACILITATOR_ERROR" };
        }
        catch
        {
            return new CheckResult { IsValid = false, InvalidReason = "FACILITATOR_ERROR" };
        }
    }

    public void Dispose() => _http.Dispose();
}
